// react-native-counter-timer/index.js
// @flow
import React from 'react';
import {
  View,
  Text,
  TouchableOpacity,
} from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import PropTypes from 'prop-types';
import type {Element as ReactElement} from 'react';

import ActivityTimerComponent from './src/containers/activityTimer/ActivityTimerComponent';

import {
  SKIP,
  REST,
  RESET,
  PRIMARY_ACTION,
  TOP,
  BOTTOM,
  MAX_TIME,
  TIMER,
  SET,
  MINUITES,
  SECONDS,
} from './src/shared/strings';
import styles, {
  containerStyleProps,
  gradientColorsRepsDefault,
  gradientColorsDefault,
  progressColorRest,
  progressColorTimer,
} from './CustomCounterTimer.styles';

import colors from './src/shared/themes/colors';
import {ResetComponent, PrimaryActionComponent} from './src/shared/components';

type CustomCounterTimerProps = {
  timer: TimerModel,
};
type CustomCounterTimerState = {
  currentRunningSet: number,
  status: ACTIVITY_STATUS.NOT_STARTED | ACTIVITY_STATUS.IN_PROGRESS | ACTIVITY_STATUS.RESTING | ACTIVITY_STATUS.COMPLETED,
  timerRunning: bool,
  restTimerRunning: bool,
};

const ACTIVITY_STATUS = {
  NOT_STARTED: 10001,
  IN_PROGRESS: 10002,
  REST: 10003,
  COMPLETED: 10004,
};
class CustomCounterTimerContainer extends React.PureComponent<CustomCounterTimerProps, CustomCounterTimerState> {
  static defaultProps: any

  constructor(props: CustomCounterTimerProps) {
    super(props);
  }

  state = {
    currentRunningSet: 0,
    status: ACTIVITY_STATUS.NOT_STARTED,
    timerRunning: false,
    timer: this.props.timer,
    name: '',
    restTimerRunning: false,

    isMuted: false,
  }

  componentWillMount() {
    const {timer} = this.props;
    if (timer) {
      const timerData = {
        activiTimeHours: timer.activiTimeHours,
        activeTimeMinutes: timer.activeTimeMinutes,
        activeTimeSeconds: timer.activeTimeSeconds,
        id: timer.id,
        name: timer.name,
        restTimeHours: timer.restTimeHours,
        restTimeMinutes: timer.restTimeMinutes,
        restTimeSeconds: timer.restTimeSeconds,
        sets: timer.sets,
      };
      this.setState({
        name: timerData.name,
      });
    }
  }

  componentWillUnmount() {
  }

  restTone = null;
  activeTone = null;

  // _timerRef: ActivityTimerComponent;
  // _restTimerRef: ActivityTimerComponent;

  handleTimeRef = (ref) => {
    this._timerRef = ref;
  };

  handleRestTimeRef = (ref) => {
    this._restTimerRef = ref;
  };

  resetTimerRef = () => {
    this._timerRef && this._timerRef.reset();
    this._restTimerRef && this._restTimerRef.reset();
  }

  resolveContainerBorderStyle = (): number | null => {
    const {status} = this.state;
    let borderStyle = {
      borderRadius: 8,
    };

    switch (status) {
    case ACTIVITY_STATUS.IN_PROGRESS:
      borderStyle = styles.borderGreen;
      break;
    case ACTIVITY_STATUS.REST:
      borderStyle = styles.borderOrange;
      break;
    }

    return borderStyle;
  }

  formatSetLabel = (): string | null => {
    const {timer: {sets}, currentRunningSet} = this.state;
    const {counterSetText, counterSetSeperatorText} = this.props;

    let result = null;

    if (sets > 1) {
      result = `${counterSetText} ${currentRunningSet + 1}${counterSetSeperatorText}${sets}`;
    }

    return result;
  }

  restSoundPlay = () => {
    if (this.restTone != null && !this.state.isMuted) {
      this.restTone.play((success) => {
        if (!success) {
          // ToastAndroid.show(ERROR_WHEN_INIT_SOUND_PLAYER, ToastAndroid.SHORT);
        }
      });
    }
  }

  activeSoundPlay = () => {
    if (this.activeTone != null && !this.state.isMuted) {
      this.activeTone.play((success) => {
        if (!success) {
          // ToastAndroid.show(ERROR_WHEN_INIT_SOUND_PLAYER, ToastAndroid.SHORT);
        }
      });
    }
  }

  moveForwardToNextSet = (): void => {
    const {currentRunningSet} = this.state;

    this.setState({
      currentRunningSet: currentRunningSet + 1,
      restTimerRunning: false,
      timerRunning: true,
      status: ACTIVITY_STATUS.IN_PROGRESS,
    });
  }

  moreSetsAvaileble = (): bool => {
    const {timer: {sets: totalSets}} = this.state;
    const {currentRunningSet} = this.state;

    return (currentRunningSet + 1) < totalSets;
  }

  restAvailableForActivity = (): bool => {
    const {timer: {restTimeMinutes, restTimeSeconds}} = this.state;

    return restTimeMinutes > 0 || restTimeSeconds > 0;
  }

  resolveMovingForwardFromRest = (): void => {
    // TODO:
    if (this.moreSetsAvaileble()) {
      // TODO:
      // Alert.alert('if (this.moreSetsAvaileble()) {');
      // There are more sets to go. Moving on to next set
      this.activeSoundPlay();
      this.moveForwardToNextSet();
    } else {
      // TODO:
      // Alert.alert('} else {');
      // This is the final rest of the activity. Let's wrap this up
      this.handleActivityCompleted();
    }
  }

  handleMuteToggled = (): void => {
    const {onMuteToggle, onUnMuteToggle} = this.props;
    const {isMuted} = this.state;

    if (!isMuted && onMuteToggle) {
      onMuteToggle();
    }

    if (isMuted && onUnMuteToggle) {
      onUnMuteToggle();
    }

    this.setState({
      isMuted: !isMuted,
    });
  }

  handleResetPressed = (): void => {
    const {onResetButtonPressed} = this.props;

    if (onResetButtonPressed) {
      onResetButtonPressed();
    }

    this.resetTimerRef();
    this.setState({
      currentRunningSet: 0,
      status: ACTIVITY_STATUS.NOT_STARTED,
      restTimerRunning: false,
      timerRunning: false,
    });
  }

  handleStartTimerPressed = (): void => {
    const {onStartTimerPressed} = this.props;
    if (onStartTimerPressed) {
      onStartTimerPressed();
    }
    this.activeSoundPlay();
    this.setState({
      status: ACTIVITY_STATUS.IN_PROGRESS,
      timerRunning: true,
    });
  }

  handleSkipSet = (): void => {
    const lastSet = !this.moreSetsAvaileble();
    let partialState = null;

    if (lastSet) {
      this.resetTimerRef();
      partialState = {status: ACTIVITY_STATUS.COMPLETED};
    } else if (!lastSet) {
      partialState = {
        status: ACTIVITY_STATUS.NOT_STARTED,
        currentRunningSet: this.state.currentRunningSet + 1,
        timerRunning: false,
        restTimerRunning: false,
      };
    } else {
      this.handleActivityCompleted();
    }

    if (partialState) {
      this.setState(partialState);
    }
  }

  handleSecondaryModePausePressed = (): void => {
    const {onSecondaryModePausePressed} = this.props;

    if (onSecondaryModePausePressed) {
      onSecondaryModePausePressed();
    }

    this.setState({
      restTimerRunning: false,
    });
  }

  handleSecondaryModeStartPressed = (): void => {
    const {onSecondaryModeStartPressed} = this.props;

    if (onSecondaryModeStartPressed) {
      onSecondaryModeStartPressed();
    }

    this.setState({
      restTimerRunning: true,
    });
  }

  handlePrimaryModePausePressed = (): void => {
    const {onPrimaryModePausePressed} = this.props;

    if (onPrimaryModePausePressed) {
      onPrimaryModePausePressed();
    }

    this.setState({
      timerRunning: false,
    });
  }

  handlePrimaryModeStartPressed = (): void => {
    const {onPrimaryModeStartPressed} = this.props;

    if (onPrimaryModeStartPressed) {
      onPrimaryModeStartPressed();
    }
    this.setState({
      timerRunning: true,
    });
  }

  handleNextSet = (): void => {
    this.activeSoundPlay();
    const partialState = {
      restTimerRunning: false,
      timerRunning: true,
      status: ACTIVITY_STATUS.IN_PROGRESS,
      currentRunningSet: this.state.currentRunningSet + 1,
    };
    this.setState(partialState);
  }

  handleActivityCompleted = (): void => {
    const {onActivityCompleted} = this.props;

    if (onActivityCompleted) {
      onActivityCompleted();
    }

    const partialState = {
      restTimerRunning: false,
      timerRunning: false,
      status: ACTIVITY_STATUS.COMPLETED,
    };
    this.setState(partialState);
  }

  handleSkipPressed = (): void => {
    const {status} = this.state;
    const {onSkipPressed} = this.props;

    if (onSkipPressed) {
      onSkipPressed();
    }

    switch (status) {
    case ACTIVITY_STATUS.NOT_STARTED:
      this.handleSkipSet();
      break;
    case ACTIVITY_STATUS.IN_PROGRESS:
      this.handleSkipSet();
      break;
    case ACTIVITY_STATUS.REST:
      this.resolveMovingForwardFromRest();
      break;
    }
  }

  handleSecondaryModeEnd = (): void => {
    const {onSecondaryModeEnd} = this.props;

    if (onSecondaryModeEnd) {
      onSecondaryModeEnd();
    }

    this.restSoundPlay();
    this.resolveMovingForwardFromRest();
  }

  handleTimeEnd = (): void => {
    const lastSet = !this.moreSetsAvaileble();
    let partialState = null;

    if (this.restAvailableForActivity()) {
      this.restSoundPlay();
      partialState = {
        status: ACTIVITY_STATUS.REST,
        restTimerRunning: true,
      };
    } else {
      if (lastSet) {
        partialState = {
          status: ACTIVITY_STATUS.COMPLETED,
          timerRunning: false,
          restTime: false,
        };
      } else {
        this.handleNextSet();
      }
    }

    if (partialState) {
      this.setState(partialState);
    }
  }

  renderMuteIcon = () => {
    return (
      <TouchableOpacity
        hitSlop={{top: 15, bottom: 15, left: 15, right: 15}}
        onPress={this.handleMuteToggled}
        style={styles.muteIconWrapper}
      >
        <Text>sss</Text>
        {/* <Image
            source={isMuted ? mutedIcon : muteIcon}
            style={styles.settingStyle}
          /> */}
      </TouchableOpacity>
    );
  }

  renderActivityTopItems = (): Array<ReactElement<any>> => {
    const {status} = this.state;
    const {showMuteElement, topItemsWrapperStyle} = this.props;

    let statusText = this.state.name.toUpperCase();
    let statusTextStyle = styles.defaultStatusTextStyle;
    const muteIcon = showMuteElement ? this.renderMuteIcon() : null;

    if (status == ACTIVITY_STATUS.REST) {
      statusText = REST;
      statusTextStyle = styles.resetTextStyle;
    } else if (status == ACTIVITY_STATUS.IN_PROGRESS) {
      // statusText = ACTIVE;
      statusTextStyle = styles.activeTextStyle;
    } else {
      statusText = this.state.name.toUpperCase();
    }

    return (
      <View style={topItemsWrapperStyle}>
        <View style={styles.settingsIconWrapper}>
          {this.props.leftUpperElement}
        </View>

        <Text style={[styles.activityStatusText, statusTextStyle]}>
          {statusText}
        </Text>
        {muteIcon}
      </View>
    );
  }

  renderPrimaryActionButton = (): ReactElement<any> => {
    const {status, restTimerRunning, timerRunning} = this.state;
    const {controllerMainPrimaryActionButtonStyle, controllerMainPrimaryActionDisabledButtonStyle, controllerMainPrimaryActionButtonElement,
      controllerMainPrimaryActionButtonTextStyle, controllerSecondPrimaryActionButtonTextStyle, controllerSecondPrimaryActionButtonElement, controllerSecondPrimaryActionButtonStyle} = this.props;

    // TODO
    const PrimaryActionWrapperStyle = {flex: 1, alignItems: 'center'};

    return (
      <View style={PrimaryActionWrapperStyle}>
        <PrimaryActionComponent
          controllerMainPrimaryActionButtonElement={controllerMainPrimaryActionButtonElement}
          controllerMainPrimaryActionButtonStyle={controllerMainPrimaryActionButtonStyle}
          controllerMainPrimaryActionButtonTextStyle={controllerMainPrimaryActionButtonTextStyle}
          controllerMainPrimaryActionDisabledButtonStyle={controllerMainPrimaryActionDisabledButtonStyle}
          controllerSecondPrimaryActionButtonElement={controllerSecondPrimaryActionButtonElement}
          controllerSecondPrimaryActionButtonStyle={controllerSecondPrimaryActionButtonStyle}
          controllerSecondPrimaryActionButtonTextStyle={controllerSecondPrimaryActionButtonTextStyle}
          onActivityCompleted={this.handleActivityCompleted}
          onPrimaryModePausePressed={this.handlePrimaryModePausePressed}
          onPrimaryModeStartPressed={this.handlePrimaryModeStartPressed}
          onSecondaryModePausePressed={this.handleSecondaryModePausePressed}
          onSecondaryModeStartPressed={this.handleSecondaryModeStartPressed}
          onStartTimerPressed={this.handleStartTimerPressed}
          restTimerRunning={restTimerRunning}
          status={status}
          timerRunning={timerRunning}
        />
      </View>
    );
  }

  renderResetButton = (): ReactElement<any> => {
    const {status, currentRunningSet} = this.state;
    const {controllerResetText, controllerResetButtonStyle, controllerDisabledResetButtonTextStyle,
      controllerDisabledResetButtonStyle, controllerResetButtonTextStyle, controllerResetButtonElement, controllerDisabledResetButtonElement} = this.props;

    return (
      <ResetComponent
        controllerDisabledResetButtonElement={controllerDisabledResetButtonElement}
        controllerDisabledResetButtonStyle={controllerDisabledResetButtonStyle}
        controllerDisabledResetButtonTextStyle={controllerDisabledResetButtonTextStyle}
        controllerResetButtonElement={controllerResetButtonElement}
        controllerResetButtonStyle={controllerResetButtonStyle}
        controllerResetButtonTextStyle={controllerResetButtonTextStyle}
        controllerResetText={controllerResetText}
        currentRunningSet={currentRunningSet}
        onResetPressed={this.handleResetPressed}
        status={status}
      />
    );
  }

  renderSkipButton = (): ReactElement<any> => {
    const {status} = this.state;
    const {controllerSkipButtonText, controllerDisabledSkipButtonTextStyle, controllerSkipButtonTextStyle,
      controllerSkipButtonStyle, controllerDisabledSkipButtonStyle, controllerSkipButtonElement, controllerDisabledSkipButtonElement} = this.props;
    const textStyles = [controllerSkipButtonTextStyle];
    const buttonStyle = [controllerSkipButtonStyle];
    const disabled = status === ACTIVITY_STATUS.COMPLETED;
    const skipButtonElement = controllerSkipButtonElement;

    if (disabled) {
      textStyles.push(styles.textDisabled);
      if (controllerDisabledSkipButtonTextStyle) {
        textStyles.push(controllerDisabledSkipButtonTextStyle);
      }
      if (controllerDisabledSkipButtonStyle) {
        buttonStyle.push(controllerDisabledSkipButtonStyle);
      }

      if (controllerDisabledSkipButtonElement) {
        skipButtonElement.push(controllerDisabledSkipButtonElement);
      }
    }

    const butttonElement = controllerSkipButtonElement ? skipButtonElement : ( <Text style={textStyles}>{`${controllerSkipButtonText}`}</Text>);

    return (
      <TouchableOpacity
        disabled={disabled}
        // hitSlop={{top: 25, bottom: 25, left: 25, right: 25}}
        onPress={this.handleSkipPressed}
        style={buttonStyle}
      >
        {butttonElement}
      </TouchableOpacity>
    );
  }

  renderActivityControls = (): ReactElement<any> => {
    const {controlsWrapperStyle, controllerButtons} = this.props;

    const resetButton = this.renderResetButton();
    const primaryActionButton = this.renderPrimaryActionButton();
    const skipButton = this.renderSkipButton();
    const content = [];

    controllerButtons.forEach((element) => {
      switch (element) {
      case RESET:
        content.push(resetButton);
        break;

      case PRIMARY_ACTION:
        content.push(primaryActionButton);
        break;

      case SKIP:
        content.push(skipButton);
        break;
      }
    });

    return (
      <View style={controlsWrapperStyle}>
        {content}
      </View>
    );
  }
  getTimerDuration = (hours, minutes, seconds): number => {
    const duration = (hours * 3600) + (minutes * 60) + seconds;

    return duration;
  }

  renderTimerContent = (): ReactElement<any> => {
    const {timer: {activeTimeMinutes, activeTimeSeconds, activiTimeHours}} = this.state;
    const {status, currentRunningSet} = this.state;
    const {progressPrimaryStatusColor, progressSecondaryStatusColor, progressVisible,
      progressBorderWidth, progressSize, progressThickness, progressUnfilledColor, progressAnimation, progressDirection, progressStyle, counterTexts,
      counterSetTextWrapperStyle, counterTimer, runInBackground} = this.props;
    const timerKey = `workoutTimer${currentRunningSet}`;
    let progressColor = null;

    const duration = this.getTimerDuration(activiTimeHours, activeTimeMinutes, activeTimeSeconds);

    switch (status) {
    case ACTIVITY_STATUS.IN_PROGRESS:
      progressColor = progressPrimaryStatusColor;

      break;
    case ACTIVITY_STATUS.REST:
      progressColor = progressSecondaryStatusColor;

      break;
    }

    return (
      <ActivityTimerComponent
        counterSetTextWrapperStyle={counterSetTextWrapperStyle}
        counterTexts={counterTexts}
        counterTimer={counterTimer}
        key={timerKey}
        label={this.formatSetLabel()}
        onCountFinish={this.handleTimeEnd}
        progressAnimation={progressAnimation}
        progressBorderWidth={progressBorderWidth}
        progressColor={progressColor}
        progressDirection={progressDirection}
        progressSize={progressSize}
        progressStyle={progressStyle}
        progressThickness={progressThickness}
        progressUnfilledColor={progressUnfilledColor}
        progressVisible={progressVisible}
        ref={this.handleTimeRef}
        runInBackground={runInBackground}
        started={this.state.timerRunning}
        timeToRun={duration}
        type={ActivityTimerComponent.COUNT_TYPE.COUNTUP}
      />
    );
  }

  renderRestContent = (): ReactElement<any> => {
    const {timer: {restTimeMinutes, restTimeSeconds, restTimeHours}} = this.state;
    const {status, currentRunningSet} = this.state;
    const {progressPrimaryStatusColor, progressSecondaryStatusColor, progressVisible,
      progressBorderWidth, progressSize, progressThickness, progressUnfilledColor, progressAnimation, progressDirection, progressStyle,
      counterTexts, counterSetTextWrapperStyle, counterTimer, runInBackground} = this.props;
    const timerKey = `restTimer${currentRunningSet}`;
    let progressColor = null;

    const restTime = this.getTimerDuration(restTimeHours, restTimeMinutes, restTimeSeconds);
    switch (status) {
    case ACTIVITY_STATUS.IN_PROGRESS:
      progressColor = progressPrimaryStatusColor;

      break;
    case ACTIVITY_STATUS.REST:
      progressColor = progressSecondaryStatusColor;

      break;
    }

    return (
      <ActivityTimerComponent
        counterSetTextWrapperStyle={counterSetTextWrapperStyle}
        counterTexts={counterTexts}
        counterTimer={counterTimer}
        key={timerKey}
        label={this.formatSetLabel()}
        onCountFinish={this.handleSecondaryModeEnd}
        progressAnimation={progressAnimation}
        progressBorderWidth={progressBorderWidth}
        progressColor={progressColor}
        progressDirection={progressDirection}
        progressSize={progressSize}
        progressStyle={progressStyle}
        progressThickness={progressThickness}
        progressUnfilledColor={progressUnfilledColor}
        progressVisible={progressVisible}
        ref={this.handleRestTimeRef}
        runInBackground={runInBackground}
        started={this.state.restTimerRunning}
        timeToRun={restTime}
        type={ActivityTimerComponent.COUNT_TYPE.COUNTUP}
      />
    );
  }

  renderCompletedPlaceholder = (): ReactElement<any> => {
    return (
      <View style={styles.completedPlaceholder}>
        <Text style={styles.completedText}>All sets completed</Text>
      </View>
    );
  }

  renderCounterContent = (): ReactElement<any> => {
    const {status} = this.state;
    const {gradientColorsRepsActive, gradientColorsRestActive, progressWrapper, gradientColorsDefault, timerStatusTextStyle, counterTimerStatusTextVisible,
      progressPrimaryStatusColor, progressSecondaryStatusColor, counterTimerDefaultStatusText, counterTimerPrimaryStatusText, counterTimerSecondaryStatusText} = this.props;
    let counter = null;
    let gradientColors = gradientColorsRepsDefault;
    let borderStyle = {
      borderRadius: 8,
    };

    const timertextStyle = [timerStatusTextStyle];
    let timerText = counterTimerDefaultStatusText;

    switch (status) {
    case ACTIVITY_STATUS.NOT_STARTED:
      gradientColors = gradientColorsDefault;
      counter = this.renderTimerContent();
      timerText = counterTimerDefaultStatusText;
      break;
    case ACTIVITY_STATUS.IN_PROGRESS:
      gradientColors = gradientColorsRepsActive;
      borderStyle = styles.borderGreen;
      counter = this.renderTimerContent();
      timertextStyle.push({color: progressPrimaryStatusColor});
      timerText = counterTimerPrimaryStatusText;
      break;
    case ACTIVITY_STATUS.REST:
      gradientColors = gradientColorsRestActive;
      borderStyle = styles.borderOrange;
      counter = this.renderRestContent();
      timertextStyle.push({color: progressSecondaryStatusColor});
      timerText = counterTimerSecondaryStatusText;
      break;
    case ACTIVITY_STATUS.COMPLETED:
      counter = this.renderTimerContent();
      break;
    }

    const timerStatus = counterTimerStatusTextVisible ? (<Text style={timertextStyle}>{timerText}</Text>) : null;

    return (
      <LinearGradient
        colors={gradientColors}
        style={[styles.contentWrapper, borderStyle]}
      >
        {/* {this.renderActivityTopItems()} */}
        {timerStatus}
        <View style={progressWrapper}>
          {counter}
        </View>
      </LinearGradient>
    );
  }

  renderBoxContainer = () => {
    const {controllerPosition} = this.props;
    const controls = this.renderActivityControls();
    const content = this.renderCounterContent();
    const elements = controllerPosition == TOP ? (
      <View
        {...containerStyleProps}
        style={[containerBorderStyle]}
      >
        {controls}
        {content}
      </View>
    ) : (
      <View
        {...containerStyleProps}
        style={[containerBorderStyle]}
      >
        {content}
        {controls}
      </View>
    );

    const containerBorderStyle = this.resolveContainerBorderStyle();

    return (
      elements
    );
  }

  render() {
    const container = this.renderBoxContainer();

    return container;
  }
}

CustomCounterTimerContainer.propTypes = {
  controllerButtons: PropTypes.any,
  controllerDisabledResetButtonElement: PropTypes.any,
  controllerDisabledResetButtonStyle: PropTypes.any,
  controllerDisabledResetButtonTextStyle: PropTypes.any,
  controllerDisabledSkipButtonElement: PropTypes.any,
  controllerDisabledSkipButtonStyle: PropTypes.any,
  controllerDisabledSkipButtonTextStyle: PropTypes.any,
  controllerMainPrimaryActionButtonElement: PropTypes.any,
  controllerMainPrimaryActionButtonStyle: PropTypes.any,
  controllerMainPrimaryActionButtonTextStyle: PropTypes.any,
  controllerMainPrimaryActionDisabledButtonStyle: PropTypes.any,
  controllerPosition: PropTypes.string,
  controllerResetButtonElement: PropTypes.any,
  controllerResetButtonStyle: PropTypes.any,
  controllerResetButtonTextStyle: PropTypes.any,
  controllerResetText: PropTypes.string,
  controllerSecondPrimaryActionButtonElement: PropTypes.any,
  controllerSecondPrimaryActionButtonStyle: PropTypes.any,
  controllerSecondPrimaryActionButtonTextStyle: PropTypes.any,
  controllerSkipButtonElement: PropTypes.any,
  controllerSkipButtonStyle: PropTypes.any,
  controllerSkipButtonText: PropTypes.string,
  controllerSkipButtonTextStyle: PropTypes.any,
  controlsWrapperStyle: PropTypes.any,

  counterSetSeperatorText: PropTypes.string,
  counterSetText: PropTypes.string,
  counterSetTextWrapperStyle: PropTypes.any,
  counterTexts: PropTypes.any,
  counterTimer: PropTypes.any,
  counterTimerDefaultStatusText: PropTypes.string,
  counterTimerPrimaryStatusText: PropTypes.string,
  counterTimerSecondaryStatusText: PropTypes.string,
  counterTimerStatusTextVisible: PropTypes.bool,

  gradientColorsDefault: PropTypes.array,
  gradientColorsRepsActive: PropTypes.array,
  gradientColorsRestActive: PropTypes.array,

  leftUpperElement: PropTypes.element,

  onActivityCompleted: PropTypes.func,
  onMuteToggle: PropTypes.func,
  onPrimaryModePausePressed: PropTypes.func,
  onPrimaryModeStartPressed: PropTypes.func,
  onResetButtonPressed: PropTypes.func,
  onSecondaryModeEnd: PropTypes.func,
  onSecondaryModePausePressed: PropTypes.func,
  onSecondaryModeStartPressed: PropTypes.func,
  onSkipPressed: PropTypes.func,
  onStartTimerPressed: PropTypes.func,
  onUnMuteToggle: PropTypes.func,

  progressAnimation: PropTypes.bool,
  progressBorderWidth: PropTypes.number,
  progressDirection: PropTypes.string,
  progressPrimaryStatusColor: PropTypes.string,
  progressSecondaryStatusColor: PropTypes.string,
  progressSize: PropTypes.number,
  progressStyle: PropTypes.any,
  progressThickness: PropTypes.number,
  progressVisible: PropTypes.bool,
  progressWrapper: PropTypes.any,
  runInBackground: PropTypes.bool,

  showMuteElement: PropTypes.bool,

  timer: PropTypes.shape({
    restTimeMinutes: PropTypes.number.isRequired,
    restTimeSeconds: PropTypes.number.isRequired,
    activeTimeMinutes: PropTypes.number.isRequired,
    activeTimeSeconds: PropTypes.number.isRequired,
    sets: PropTypes.number.isRequired,
  }).isRequired,
  timerStatusTextStyle: PropTypes.any,
  topItemsWrapperStyle: PropTypes.any,
};

CustomCounterTimerContainer.defaultProps = {
  leftUpperElement: null,
  onMuteToggle: null,
  onPrimaryModePausePressed: null,
  onStartTimerPressed: null,
  onResetButtonPressed: null,
  onSecondaryModeStartPressed: null,
  onSecondaryModePausePressed: null,
  onPrimaryModeStartPressed: null,
  onSecondaryModeEnd: null,
  onActivityCompleted: null,
  onSkipPressed: null,
  onUnMuteToggle: null,
  progressSecondaryStatusColor: progressColorRest,
  progressPrimaryStatusColor: progressColorTimer,
  gradientColorsDefault: gradientColorsDefault,
  gradientColorsRepsActive: gradientColorsDefault,
  gradientColorsRestActive: gradientColorsDefault,
  counterSetText: 'Set',
  showMuteElement: true,
  progressVisible: true,
  controlsWrapperStyle: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    paddingBottom: 10,
    paddingTop: 10,
  },
  progressWrapper: {
    paddingTop: 10,
  },
  topItemsWrapperStyle: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignSelf: 'stretch',
  },
  controllerButtons: [PRIMARY_ACTION],
  controllerPosition: BOTTOM,
  controllerResetText: RESET,
  controllerResetButtonStyle: {
    paddingTop: 10,
    alignItems: 'center',
    flex: 1,
  },
  controllerDisabledResetButtonTextStyle: null,
  controllerResetButtonTextStyle: {
    color: colors.background.greenCrock,
    fontSize: 16,
  },
  controllerSkipButtonText: SKIP,
  controllerDisabledSkipButtonTextStyle: null,
  controllerSkipButtonTextStyle: {
    color: colors.background.greenCrock,
    fontSize: 16,
  },
  controllerSkipButtonStyle: {
    paddingTop: 10,
    alignItems: 'center',
    flex: 1,
  },
  controllerDisabledSkipButtonStyle: null,
  controllerDisabledResetButtonStyle: null,
  controllerMainPrimaryActionButtonStyle: null,
  controllerMainPrimaryActionDisabledButtonStyle: null,
  controllerSecondPrimaryActionButtonStyle: null,
  counterTimerDefaultStatusText: 'Default',
  counterTimerPrimaryStatusText: 'primary',
  counterTimerSecondaryStatusText: 'secondary',
  controllerMainPrimaryActionButtonElement: null,
  controllerSecondPrimaryActionButtonTextStyle: null,
  controllerSecondPrimaryActionButtonElement: null,

  progressBorderWidth: 0,
  progressSize: 140,
  progressThickness: 4,
  progressAnimation: true,
  progressDirection: 'counter-clockwise',
  progressStyle: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  counterTexts: [MAX_TIME, TIMER, SET],
  counterSetSeperatorText: '/',
  counterSetTextWrapperStyle: {
    fontSize: 20,
    color: colors.background.black,
  },
  counterTimer: [MINUITES, SECONDS],
  timerStatusTextStyle: {
    color: colors.background.black,
  },

  counterTimerStatusTextVisible: null,

  controllerMainPrimaryActionButtonTextStyle: null,
  controllerResetButtonElement: null,
  controllerDisabledResetButtonElement: null,
  controllerSkipButtonElement: null,
  controllerDisabledSkipButtonElement: null,

  runInBackground: true,
};

export default CustomCounterTimerContainer;
