// react-native-counter-timer/src/shared/themes/colors.js
// @flow
const black = '#000000';
const white = '#ffffff';
/**
 * App colors
 * @type {Object}
 */
export default {
  /**
   * Brand colors
   * @type {Object}
   */
  brand: {
    appleGreen: '#7ed321',
  },

  /**
   * Background colors
   * @type {Object}
   */
  background: {
    container: '#F8F8F8',
    white,
    black,
    purpleishBlue: '#3D3BEE',
    blueCrock: '#0086C6',
    transparentblueCrock: '#0086C650',
    buff: '#FEF697',
    eggShellBlue: '#C2FFF1',
    jetBlack: '#262626',
    yellow: '#CCCC00',
    cerulean: '#0086C6',
    coolGray: '#AFB8C1',
    offWhite: '#F8F8F8',
    kermitGreen: '#63C401',
    orangeYellow: '#FFA900',
    barbiePink: '#FF34AA',
    steel: '#8E8E93',
    lightSteel: '#D4D6D6',
    gray: '#808080',
    grayTransparent: '#80808080',
    red: 'red',
    activeOrange: '#FF9800',
    restGreen: '#AEEA00',
    lightOrange: '#FFF9EC',
    lightGreen: '#FDFFF4',
    activeLightOrange: '#F9E9D0',
    activeLightGreen: '#FDFEF3',
    darkGray: '#2f4f4f',
    ligtBlue: '#D9EEF8',
    disabledBlue: '#81C4E3',
    weightColor: '#757575',
    lightBlack: '#262626',
    disableBlueCrock: '#90CDEA',
    lightBlueCrock: '#D6EEF9',
    transprentSpinner: 'rgba(0, 0, 0, 0.9)',
    memberAreaBlue: 'rgba(0, 134, 198, 0.15)',
    greyLinear: 'rgba(255, 255, 255, 0)',
    darkLinear: 'rgba(41, 41, 41, 0.54)',
    topLinear: 'rgba(0, 0, 0, 0.54)',
    darkYellow: '#FEEA33',
    lightYellow: '#FEEB03',
    sunflowerYellow: '#FFD604',
    greyishBrown: '#555555',
    whiteTwo: '#e4e5d5',
    whiteFour: '#F4F4F4',
    whiteFive: '#E4E4E4',
    warmGrey: '#8D8D8D',
    transparentBlack: '#00000050',
    transparent: 'rgba(255, 255, 255, 0)',
    transparentWhite: '#ffffff95',
    appGray: '#E0E0E0',
    greenCrock: '#116466',
    disbaledGreenCrock: '#649A9B',
    transparent: 'transparent',
  },

  tabBar: {
    backgroundColor: 'white',
    iconColor: '#000000',
    whiteTwo: '#ECECEC',
    black,
  },

  /**
   * Font colors
   * @type {Object}
   */
  font: {
    primary: '#000000',
    secondary: '#333333',
    white,
    error: '#eb2500',
    red: '#ff0000',
    coolGray: '#AFB8C1',
    purpleishBlue: '#3D3BEE',
    purpleishBlueDisabled: '#3D3BEE50',
    black,
    warmGrey: '#8D8D8D',
    warmGreyTwo: '#828282',
    darkGray: '#565656',
    lightGray: '#6e6e6e',
    yellow: '#CCCC00',
    darkYellow: '#FEEA33',
    greyishBrown: '#555555',
    darkSkyBlue: '#409bF1',
    BlueCrock: '#0086C6',
    weightColor: '#757575',
    blackTwo: '#262626',
  },
  divider: {
    mainDivider: '#6C7A89',
    itemDivider: '#e1e7ed',
  },
  profilePage: {
    divider: {
      mainDivider: '#6C7A89',
    },
    trialBorder: '#ddd',
    trialShadow: '#000',
  },
  buttons: {
    buttonOverlay: '#808080',
    play: '#A9A9A9',
  },
  timer: {
    lemonGreen: '#aeea00',
    tangerine: '#ff9800',
    cerulean1: '#fff3e0',
    cerulean: '#f9fbe7',
  },

  activity: {
    borderDefault: '#E0E0E0',
    borderOrange: '#E85A4F',
    borderGreen: '#AEEA00',
    backgroundOrange: '#FFF9EC',
    backgroundGreen: '#FDFFF4',
    borderBlue: '#31708E',
    white,
    black,

    progressBackground: '#E8E8E8',
  },

  transparent: 'transparent',
};
