// react-native-counter-timer/src/containers/activityTimer/ActivityTimer.styles.js
// @flow
import {StyleSheet} from 'react-native';

// import {typeFace} from '@crock:theme/base';
import colors from '../../shared/themes/colors';

const style = StyleSheet.create({
  container: {
  },
  timeTextWrapper: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    width: '100%',
  },
  totalTimeText: {
    // ...typeFace,
    // flex: 1,
    fontSize: 16,
    color: colors.font.weightColor,
  },
  totalTimeColon: {
    // ...typeFace,
    fontSize: 16,
  },
  elapsedTimeText: {
    // ...typeFace,
    fontSize: 38,
    color: colors.background.black,
    width: 50,
    textAlign: 'center',
  },
  timeMinutesRightAlignText: {
    // justifyContent: 'flex-end',
    textAlign: 'right',
    alignSelf: 'stretch',
  },
  elapsedTimeSeconds: {},
  elapsedTimeColon: {
    // ...typeFace,
    fontSize: 38,
    color: colors.background.black,
    marginBottom: 8,
  },
  labelText: {
    // ...typeFace,
    fontSize: 20,
    color: colors.background.black,
    // alignItems: 'center',
    // justifyContent: 'space-between',
  },
});

export default style;

export const progressCircleDefaultStyleProps = {
  borderWidth: 0,
  size: 140,
  thickness: 4,
  unfilledColor: colors.activity.progressBackground,
};
